File: cpp/recursion_depth_first_search.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

void depthFirstPrint(unordered_map<char, vector<char>>& graph, char source) {
    cout << source << " ";
    for (char neighbor : graph[source]) {
        depthFirstPrint(graph, neighbor);
    }
}

int main() {
    unordered_map<char, vector<char>> graph;
    graph['a'] = {'c', 'b'};
    graph['b'] = {'d'};
    graph['c'] = {'e'};
    graph['d'] = {'f'};
    graph['e'] = {};
    graph['f'] = {};

    depthFirstPrint(graph, 'a');  // Output: a c e b d f

    return 0;
}

File: new_py/recursion_depth_first_search.py
def depth_first_print(graph, source):
    print(source)
    for neighbor in graph[source]:
        depth_first_print(graph, neighbor)

graph = {
    'a': ['c', 'b'],
    'b': ['d'],
    'c': ['e'],
    'd': ['f'],
    'e': [],
    'f': []
}

depth_first_print(graph, 'a')  # Output: a c e b d f
